src/mxcp/server/interfaces/server/mcp.py

```diff
@@ -601,177 +603,168 @@ def _initialize_telemetry(self) -> None:
+                    f"Timeout waiting for requests after {timeout}s (active: {current_count}), proceeding with reload"
                 )
                 break
 
-            # Wait with timeout
-            self.drain_complete.wait(timeout=0.5)
+            # Sleep briefly to avoid busy waiting
+            time.sleep(0.1)
 
         logger.info("All requests drained, acquiring exclusive lock...")
         with self.execution_lock:
-            logger.info(f"Exclusive lock acquired. Starting {operation_name}...")
+            logger.info("Exclusive lock acquired. Starting full system reload...")
             try:
-                # Shutdown runtime components
+                # Shutdown everything
                 self._shutdown_runtime_components()
-
-                # Run the specific reload logic
-                reload_func()
-
-                # Recreate runtime components
+                
+                # Reload configuration
+                logger.info("Reloading external configuration references...")
+                new_site_config, new_user_config = self.ref_tracker.resolve_all()
+                self.site_config = new_site_config
+                self.user_config = new_user_config
+                logger.info("External references reloaded successfully")
+                
+                # Recreate everything (including Python init hooks)
                 self._initialize_runtime_components()
-
-                logger.info(f"{operation_name} completed successfully.")
-
-                # Record success metric
-                record_counter(
-                    f"mxcp.{metric_name}",
-                    attributes={"status": "success", "profile": self.profile_name},
-                    description=f"Total {operation_name} operations",
-                )
-
+                
+                logger.info("Full system reload completed successfully.")
+                
             except Exception as e:
-                logger.error(f"Failed {operation_name}: {e}", exc_info=True)
-
-                # Record failed metric
-                record_counter(
-                    f"mxcp.{metric_name}",
-                    attributes={"status": "error", "profile": self.profile_name},
-                    description=f"Total {operation_name} operations",
+                logger.error(f"Failed configuration reload: {e}", exc_info=True)
+                logger.error(
+                    "Server state may be inconsistent due to reload failure. Consider restarting."
                 )
                 raise
             finally:
                 # Always clear draining flag
                 self.draining = False
-                logger.info(f"{operation_name} process completed, normal operations resumed")
+                logger.info("Configuration reload process completed, normal operations resumed")
 
     def reload_configuration(self) -> None:
```

Comment:
`reload_configuration` is the SIGHUP handler.

If two SIGHUP signals are received close to each other, there would be two threads executing `_drain_all_requests_and_reload`. The handler seems to spawn a thread, two SIGHUP would spawn two threads.

Both execute that code:
```
        # Step 1: Start draining
        self.draining = True
        self.drain_complete.clear()
```
then wait for active requests to finish.

Say one runs first and performs the reload, eventually setting `draining` to `False`.

The second one would then execute the reload in a moment where `draining` is set to `False`, meaning other regular requests could be executing?